src/main/java/com/marklogic/kafka/connect/source/XmlRowBatcherSupplier.java

```diff
@@ -0,0 +1,101 @@
+    private final DocumentBuilderFactory factory;
+
+    XmlRowBatcherSupplier(DataMovementManager dataMovementManager, List<SourceRecord> sourceRecords, Map<String, Object> parsedConfig) {
+        this.dataMovementManager = dataMovementManager;
+        this.sourceRecords = sourceRecords;
+        this.parsedConfig = parsedConfig;
+        this.topic = (String) parsedConfig.get(MarkLogicSourceConfig.TOPIC);
+        factory = DocumentBuilderFactory.newInstance();
+    }
+
+    public RowBatcher<Document> get() {
+        ContentHandle<Document> domHandle = new DOMHandle().withMimetype("application/xml");
+        RowBatcher<Document> rowBatcher =  dataMovementManager.newRowBatcher(domHandle);
+        configureRowBatcher(parsedConfig, rowBatcher);
+        rowBatcher.onSuccess(event -> onSuccessHandler(event, sourceRecords));
+        return rowBatcher;
+    }
+
+    protected void onSuccessHandler(RowBatchSuccessListener.RowBatchResponseEvent<?> event, List<SourceRecord> newSourceRecords) {
+        @SuppressWarnings("unchecked") Document rowsDoc = ((RowBatchSuccessListener.RowBatchResponseEvent<Document>) event).getRowsDoc();
+
+        Element docElement = rowsDoc.getDocumentElement();
+        Node columnsNode = docElement.getElementsByTagNameNS(TABLE_NS_URI, "columns").item(0);
+        NodeList rows = docElement.getElementsByTagNameNS(TABLE_NS_URI, "row");
+
+        for (int i=0; i < rows.getLength(); i++) {
+            try {
+                Document newDocument = convertRowNodeToNewDocument(columnsNode, rows.item(i));
+                SourceRecord newRecord = new SourceRecord(null, null, topic, null, documentToString(newDocument));
+                newSourceRecords.add(newRecord);
+            } catch (Exception ex) {
+                logger.error("Failed to create or add SourceRecord from result row; cause: " + ex.getMessage());
+                logger.error("Failed result row: " + rows.item(i).toString());
+            }
+        }
+    }
+
+    private Document convertRowNodeToNewDocument(Node columnsNode, Node row)
+        throws TransformerException, ParserConfigurationException {
+        DocumentBuilder builder = factory.newDocumentBuilder();
+        Document newDocument = builder.newDocument();
+        Element newRoot = newDocument.createElement("Record");
+        newDocument.appendChild(newRoot);
+
+        Node newRow = row.cloneNode(true);
+        newDocument.adoptNode(newRow);
+
+        Node newColumns = columnsNode.cloneNode(true);
+        newDocument.adoptNode(newColumns);
+
+        newRoot.appendChild(newColumns);
+        newRoot.appendChild(newRow);
+        logger.debug("XML root: \n" + documentToString(newDocument));
+        return newDocument;
+    }
+
+    private static String documentToString(Document newDoc) throws TransformerException {
+        DOMSource domSource = new DOMSource(newDoc);
+        Transformer transformer = TransformerFactory.newInstance().newTransformer();
```

Comment:
Should stash this Transformer as a class attribute during the constructor, there's a cost to creating both it and the TransformerFactory. Though - check the javadocs on whether a Transformer is threadsafe - it may be that we need to stash the TransformerFactory (which I'm almost certain is threadsafe) and create a new Transformer every time.

Reply:
From the javadocs:
An object of this class may not be used in multiple threads running concurrently. Different Transformers may be used concurrently by different threads.

Reply:
Since neither of those classes are thread safe, I moved the instantiations to the beginning of the onSuccess handler.